gpu/mig.sh

```diff
@@ -0,0 +1,349 @@
+    execute_with_retries "apt-get install -y -q --no-install-recommends cuda-drivers-460"
+    execute_with_retries "apt-get install -y -q --no-install-recommends ${cuda_package}"
+  elif [[ ${OS_NAME} == centos ]]; then
+    execute_with_retries "dnf config-manager --add-repo ${NVIDIA_CENTOS_REPOSITORY_URL}"
+    execute_with_retries "dnf clean all"
+    execute_with_retries "dnf -y -q module install nvidia-driver:460-dkms"
+    execute_with_retries "dnf -y -q install cuda-${CUDA_VERSION//./-}"
+  else
+    echo "Unsupported OS: '${OS_NAME}'"
+    exit 1
+  fi
+  ldconfig
+  echo "NVIDIA GPU driver provided by NVIDIA was installed successfully"
+}
+
+# Collects 'gpu_utilization' and 'gpu_memory_utilization' metrics
+function install_gpu_agent() {
+  if ! command -v pip; then
+    execute_with_retries "apt-get install -y -q python-pip"
+  fi
+  local install_dir=/opt/gpu-utilization-agent
+  mkdir "${install_dir}"
+  curl -fsSL --retry-connrefused --retry 10 --retry-max-time 30 \
+    "${GPU_AGENT_REPO_URL}/requirements.txt" -o "${install_dir}/requirements.txt"
+  curl -fsSL --retry-connrefused --retry 10 --retry-max-time 30 \
+    "${GPU_AGENT_REPO_URL}/report_gpu_metrics.py" -o "${install_dir}/report_gpu_metrics.py"
+  pip install -r "${install_dir}/requirements.txt"
+
+  # Generate GPU service.
+  cat <<EOF >/lib/systemd/system/gpu-utilization-agent.service
+[Unit]
+Description=GPU Utilization Metric Agent
+
+[Service]
+Type=simple
+PIDFile=/run/gpu_agent.pid
+ExecStart=/bin/bash --login -c 'python "${install_dir}/report_gpu_metrics.py"'
+User=root
+Group=root
+WorkingDirectory=/
+Restart=always
+
+[Install]
+WantedBy=multi-user.target
+EOF
+  # Reload systemd manager configuration
+  systemctl daemon-reload
+  # Enable gpu-utilization-agent service
+  systemctl --no-reload --now enable gpu-utilization-agent.service
+}
+
+function enable_mig() {
+  nvidia-smi -mig 1
+}
+
+function configure_mig_cgi() {
+  if (/usr/share/google/get_metadata_value attributes/MIG_CGI); then
+    META_MIG_CGI_VALUE=$(/usr/share/google/get_metadata_value attributes/MIG_CGI)
+    nvidia-smi mig -cgi $META_MIG_CGI_VALUE -C
+  else
+    # Dataproc only supports A100's right now split in 2 if not specified
+    nvidia-smi mig -cgi 9,9  -C
+  fi
+}
+
+function main() {
+  if [[ ${OS_NAME} != debian ]] && [[ ${OS_NAME} != ubuntu ]] && [[ ${OS_NAME} != centos ]]; then
+    echo "Unsupported OS: '${OS_NAME}'"
+    exit 1
+  fi
+
+  # default MIG to on when this script is used
+  META_MIG_VALUE=1
+  if (/usr/share/google/get_metadata_value attributes/ENABLE_MIG); then
+    META_MIG_VALUE=$(/usr/share/google/get_metadata_value attributes/ENABLE_MIG)
+  fi
+  if (lspci | grep -q NVIDIA); then
+    if [[ $META_MIG_VALUE -ne 0 ]]; then
+      # check to see if we already enabled mig mode and rebooted so we don't end
+      # up in infinite reboot loop
+      NUM_GPUS_WITH_DIFF_MIG_MODES=`/usr/bin/nvidia-smi --query-gpu=mig.mode.current --format=csv,noheader | uniq | wc -l`
```

Comment:
I tried the script, but seems the start-up script failed here because the gpu driver is not installed yet:
```
Apr 19 21:48:58 test-spark-byhao2-w-0 GCEMetadataScripts[2534]: 2022/04/19 21:48:58 GCEMetadataScripts: startup-script-url: ++ /usr/bin/nvidia-smi --query-gpu=mig.mode.current --format=csv,noheader
Apr 19 21:48:58 test-spark-byhao2-w-0 GCEMetadataScripts[2534]: 2022/04/19 21:48:58 GCEMetadataScripts: startup-script-url: ++ uniq
Apr 19 21:48:58 test-spark-byhao2-w-0 GCEMetadataScripts[2534]: 2022/04/19 21:48:58 GCEMetadataScripts: startup-script-url: /tmp/metadata-scripts448357951/startup-script-url: line 282: /usr/bin/nvidia-smi: No such file or directory
Apr 19 21:48:58 test-spark-byhao2-w-0 GCEMetadataScripts[2534]: 2022/04/19 21:48:58 GCEMetadataScripts: startup-script-url: ++ wc -l
Apr 19 21:48:58 test-spark-byhao2-w-0 GCEMetadataScripts[2534]: 2022/04/19 21:48:58 GCEMetadataScripts: startup-script-url: + NUM_GPUS_WITH_DIFF_MIG_MODES=0
Apr 19 21:48:58 test-spark-byhao2-w-0 GCEMetadataScripts[2534]: 2022/04/19 21:48:58 GCEMetadataScripts: startup-script-url exit status 127
``` 

Not sure if i am doing anything wrong here?

Reply:
The same command works for debian10 image. From syslog, it seems the debian10 can tolerant the "command not found" error and continue the rest of the steps of the startup script:
```
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 google_metadata_script_runner[837]: startup-script-url: + lspci
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 google_metadata_script_runner[837]: startup-script-url: /tmp/metadata-scripts336805522/startup-script-url: line 278: lspci: command not found
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 google_metadata_script_runner[837]: startup-script-url: + grep -q NVIDIA
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 systemd[1]: man-db.service: Succeeded.
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 systemd[1]: Started Daily man-db regeneration.
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 google_metadata_script_runner[837]: startup-script-url: + [[ debian == debian ]]
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 google_metadata_script_runner[837]: startup-script-url: + export DEBIAN_FRONTEND=noninteractive
```

Reply:
Finally I figured out why:
For debian10 : It could not find "lspci" firstly,so it did not execute the commands inside the "if-else" block to trigger the "nvidia-smi". So it continues the rest of the bash script -- "mig.sh".

For ubuntu18: It could find "lspci" so it tried to run "nvidia-smi" but triggered the "nvidia-smi not found" error and as a result it exited the whole bash script because we used the variable `NUM_GPUS_WITH_DIFF_MIG_MODES` to take the output of the `nvidia-smi`.

Reply:
ok, thanks, I'll look into it, lspci code was all their before so assumed it worked, I'll test on on both os

Reply:
I tested the new script and it works for ubuntu now.